olingo/src/main/java/org/teiid/olingo/service/LocalClient.java

```diff
@@ -236,12 +257,66 @@ public void executeSQL(Query query, List<SQLParameter> parameters,
             }
             getCount = false; // the URL might have $count=true, but ignore it.
         }
+        
+        int count = 0;
+        int expectedEnd = 0;
+        
+        if (!getCount && cache) {
+            //to prevent long initial requests, we want to
+            //only work over windows of at most 32 pages
+            int resultWindow = pageSize<<5;
+            int windows = 0;
+            
+            if (toSkip != null) {
+                windows = toSkip/resultWindow;
+                toSkip = toSkip%resultWindow;
+                count = windows*resultWindow;
+            }
+            expectedEnd = count + resultWindow;
+            
+            //it's easier to just overlay an order by than to duplicate the logic from the engine that can combine limits
+            //TODO: apply this page in the prepared statement
+            if (query.getLimit() != null) {
+                query = new Query(
+                        new Select(Arrays.asList(new MultipleElementSymbol())),
+                        new From(Arrays
+                                .asList(new SubqueryFromClause("x", query))), //$NON-NLS-1$
+                        null, null, null);
+            }
+            query.setLimit(new Limit(count==0?null:new Constant(count), new Constant(resultWindow)));
+        }
+        
+        if (cache) {
+            CacheHint hint = new CacheHint();
+            hint.setTtl(getCacheTime());
+            hint.setScope(CacheDirective.Scope.USER);
+            query.setCacheHint(hint);
+        }
+        
         String sql = query.toString();
         if (cache && !Boolean.valueOf(conn.getExecutionProperty(ExecutionProperties.RESULT_SET_CACHE_MODE))) {
             sql += " /* "+ sessionId +" */"; //$NON-NLS-1$ //$NON-NLS-2$
         }
         LogManager.logDetail(LogConstants.CTX_ODATA, "Teiid-Query:",sql); //$NON-NLS-1$
         
+        if (cache) {
+            this.loadingKey = Arrays.asList(sql, parameters);
+            Future<?> future = loading.get(this.loadingKey);
+            if (future != null) {
+                try {
+                    future.get(120, TimeUnit.SECONDS);
```

Comment:
So there is hard limit of 2 minutes for query execution, this is common with browser-based clients, but should we make this user configurable perhaps?

Reply:
Two minutes is almost guaranteed to see a timeout somewhere - haproxy or another gateway. SAPUI references default timeouts of 30 and 60 seconds.  Note this will only occur on the second page, or if the same user is requesting the same set of pages at the same node as is already being loaded.  It of course could be configurable.  I'll also change it to a 503 service unavailable code.

Reply:
then do not worry about user configuration.